New dependency resolver does not properly handle missing package versions when using CPM

### NuGet Product Used

dotnet.exe, MSBuild.exe

### Product Version

6.12

### Worked before?

Yes

### Impact

It's more difficult to complete my work

### Repro Steps & Context

If you have two projects where one references the other and you're missing a PackageVersion for CPM, then the new dependency resolver does not handle this correctly.

```
ProjectA -> ProjectB -> PackageX (missing a PackageVersion)
```


During restore of each project, we verify that CPM is configured properly.  For ProjectB, we correctly identify this.  When resolving the graph of ProjectA, none of its package references are missing any versions.  I'm not sure how the legacy dependency resolver handles this.  Maybe it has to do with ordering?

### Verbose Logs

```shell
Directory.Packages.props

<Project>
  <PropertyGroup>
    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
  </PropertyGroup>
</Project>


ClassLibrary1.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <ProjectReference Include="..\ClassLibrary2\ClassLibrary2.csproj" />
  </ItemGroup>
</Project>
```

ClassLibrary2.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="EnvironmentAbstractions" />
  </ItemGroup>
</Project>

```


New resolver output:
```
  ClassLibrary2.csproj : error NU1008: Projects that use central package version management should not define the version on the PackageReference items but on the PackageVersion items: EnvironmentAbstractions.
  ClassLibrary1.csproj : error NU1101: Unable to find package EnvironmentAbstractions. No packages exist with this id in source(s): nuget.org 
```

Legacy resolver output:
```
  ClassLibrary2.csproj : error NU1008: Projects that use central package version management should not define the version on the PackageReference items but on the PackageVersion items: EnvironmentAbstractions.
```
```

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs
@@ -205,12 +205,6 @@ namespace NuGet.Commands
                     IReadOnlyDictionary<LibraryDependencyIndex, VersionRange> currentOverrides = importRefItem.VersionOverrides!;
                     bool directPackageReferenceFromRootProject = importRefItem.IsDirectPackageReferenceFromRootProject;
 
-                    // Packages with missing versions should not be added to the graph
-                    if (currentRef.LibraryRange.VersionRange == null)
-                    {
-                        continue;
-                    }
-
                     if (!findLibraryEntryCache.TryGetValue(currentRefRangeIndex, out Task<FindLibraryEntryResult>? refItemTask))
                     {
                         Debug.Fail("This should not happen");
@@ -656,6 +650,12 @@ namespace NuGet.Commands
                     for (int i = 0; i < refItemResult.Item.Data.Dependencies.Count; i++)
                     {
                         var dep = refItemResult.Item.Data.Dependencies[i];
+                        // Packages with missing versions should not be added to the graph
+                        if (dep.LibraryRange.VersionRange == null)
+                        {
+                            continue;
+                        }
+
                         LibraryDependencyIndex depIndex = refItemResult.GetDependencyIndexForDependency(i);
                         if ((dep.SuppressParent == LibraryIncludeFlags.All) && (importRefItem.LibraryDependencyIndex != rootProjectRefItem.LibraryDependencyIndex))
                         {
@@ -724,8 +724,8 @@ namespace NuGet.Commands
                         LibraryDependency dep = refItemResult.Item.Data.Dependencies[i];
                         LibraryDependencyIndex depIndex = refItemResult.GetDependencyIndexForDependency(i);
 
-                        //Suppress this node
-                        if (!importRefItem.IsCentrallyPinnedTransitivePackage && suppressions!.Contains(depIndex))
+                        // Skip this node if the VersionRange is null or if its not transitively pinned and PrivateAssets=All
+                        if (dep.LibraryRange.VersionRange == null || (!importRefItem.IsCentrallyPinnedTransitivePackage && suppressions!.Contains(depIndex)))
                         {
                             continue;
                         }
@@ -918,6 +918,12 @@ namespace NuGet.Commands
                         for (int i = 0; i < node.Item.Data.Dependencies.Count; i++)
                         {
                             var dep = node.Item.Data.Dependencies[i];
+
+                            if (dep.LibraryRange.VersionRange == null)
+                            {
+                                continue;
+                            }
+
                             if (StringComparer.OrdinalIgnoreCase.Equals(dep.Name, node.Item.Key.Name) || StringComparer.OrdinalIgnoreCase.Equals(dep.Name, rootGraphNode.Key.Name))
                             {
                                 // Cycle

--- a/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs
+++ b/src/NuGet.Core/NuGet.Commands/RestoreCommand/DependencyGraphResolver.cs
@@ -205,6 +205,12 @@ namespace NuGet.Commands
                     IReadOnlyDictionary<LibraryDependencyIndex, VersionRange> currentOverrides = importRefItem.VersionOverrides!;
                     bool directPackageReferenceFromRootProject = importRefItem.IsDirectPackageReferenceFromRootProject;
 
+                    // Packages with missing versions should not be added to the graph
+                    if (currentRef.LibraryRange.VersionRange == null)
+                    {
+                        continue;
+                    }
+
                     if (!findLibraryEntryCache.TryGetValue(currentRefRangeIndex, out Task<FindLibraryEntryResult>? refItemTask))
                     {
                         Debug.Fail("This should not happen");

--- a/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/RestoreCommand_AlgorithmEquivalencyTests.cs
+++ b/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/RestoreCommand_AlgorithmEquivalencyTests.cs
@@ -884,6 +884,69 @@ namespace NuGet.Commands.FuncTest
             result.LockFile.Targets[0].Libraries[2].Name.Should().Be("X");
         }
 
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        // Project 1 -> Project 2 -> a (null)
+        public async Task RestoreCommand_WithNullPackageVersion_AndRaisesErrorForOneProjectAndNotTheOther(bool isCPMEnabled)
+        {
+            // Arrange
+            using var pathContext = new SimpleTestPathContext();
+            var packageA = new SimpleTestPackageContext("a", "1.0.0");
+
+            await SimpleTestPackageUtility.CreateFolderFeedV3Async(
+                pathContext.PackageSource,
+                PackageSaveMode.Defaultv3,
+                packageA);
+
+            var project2spec = ProjectTestHelpers.GetPackageSpec("Project2",
+                pathContext.SolutionRoot,
+                framework: "net472");
+
+            project2spec.TargetFrameworks[0].Dependencies.Add(new LibraryDependency(
+                new LibraryRange(
+                    "a",
+                    versionRange: isCPMEnabled ? null : VersionRange.All,
+                    LibraryDependencyTarget.PackageProjectExternal)));
+
+            var project1spec = ProjectTestHelpers.GetPackageSpec("Project1",
+                pathContext.SolutionRoot,
+                framework: "net472")
+                .WithTestProjectReference(project2spec);
+
+            project1spec.RestoreMetadata.CentralPackageVersionsEnabled = isCPMEnabled;
+            project2spec.RestoreMetadata.CentralPackageVersionsEnabled = isCPMEnabled;
+
+            // Act & Assert
+            (var result, _) = await ValidateRestoreAlgorithmEquivalency(pathContext, project1spec, project2spec);
+
+            if (isCPMEnabled)
+            {
+                // Additional assert
+                result.Success.Should().BeTrue();
+                result.LogMessages.Should().BeEmpty();
+
+                result.LockFile.Targets.Should().HaveCount(1);
+                result.LockFile.Targets[0].Libraries.Should().HaveCount(1);
+                result.LockFile.Targets[0].Libraries[0].Name.Should().Be("Project2");
+                result.LockFile.Targets[0].Libraries[0].Version.Should().Be(new NuGetVersion("1.0.0"));
+            }
+            else
+            {
+
+                result.Success.Should().BeTrue();
+                result.LogMessages.Select(e => e.Code).Should().BeEquivalentTo([NuGetLogCode.NU1602]);
+
+                result.LockFile.Targets.Should().HaveCount(1);
+                result.LockFile.Targets[0].Libraries.Should().HaveCount(2);
+                result.LockFile.Targets[0].Libraries[0].Name.Should().Be("a");
+                result.LockFile.Targets[0].Libraries[0].Version.Should().Be(new NuGetVersion("1.0.0"));
+
+                result.LockFile.Targets[0].Libraries[1].Name.Should().Be("Project2");
+                result.LockFile.Targets[0].Libraries[1].Version.Should().Be(new NuGetVersion("1.0.0"));
+            }
+        }
+
         // Here's why package driven dependencies should flow.
         // Say we have P1 -> P2 -> P3 -> A 1.0.0 -> B 2.0.0
         //                            -> B 1.5.0

